src/sync/impl/sync_metadata.cpp

```diff
@@ -189,110 +212,159 @@ bool SyncMetadataManager::delete_metadata_action(const std::string& original_nam
     return true;
 }
 
-// MARK: - Sync user metadata
-
-SyncUserMetadata::SyncUserMetadata(Schema schema, SharedRealm realm, RowExpr row)
-: m_invalid(row.get_bool(schema.idx_marked_for_removal))
-, m_schema(std::move(schema))
-, m_realm(std::move(realm))
-, m_row(row)
-{ }
-
-SyncUserMetadata::SyncUserMetadata(const SyncMetadataManager& manager, std::string identity, bool make_if_absent)
-: m_schema(manager.m_user_schema)
+util::Optional<SyncUserMetadata> SyncMetadataManager::get_or_make_user_metadata(const std::string& identity,
+                                                                                const std::string& url,
+                                                                                bool make_if_absent) const
 {
     // Open the Realm.
-    m_realm = Realm::get_shared_realm(manager.get_configuration());
+    auto realm = Realm::get_shared_realm(m_metadata_config);
+    auto& schema = m_user_schema;
 
     // Retrieve or create the row for this object.
-    TableRef table = ObjectStore::table_for_object_type(m_realm->read_group(), c_sync_userMetadata);
-    size_t row_idx = table->find_first_string(m_schema.idx_identity, identity);
-    if (row_idx == not_found) {
+    TableRef table = ObjectStore::table_for_object_type(realm->read_group(), c_sync_userMetadata);
+    Query query = table->where().equal(schema.idx_identity, identity).equal(schema.idx_auth_server_url, url);
+    Results results(realm, std::move(query));
+    REALM_ASSERT_DEBUG(results.size() < 2);
+    auto maybe_row_expr = results.first();
+    size_t row_idx = not_found;
+
+    if (!maybe_row_expr) {
         if (!make_if_absent) {
-            m_invalid = true;
-            m_realm = nullptr;
-            return;
+            return none;
         }
-        m_realm->begin_transaction();
-        row_idx = table->find_first_string(m_schema.idx_identity, identity);
-        if (row_idx == not_found) {
+        realm->begin_transaction();
+        // Check the results again.
+        auto maybe_row_expr = results.first();
+        if (!maybe_row_expr) {
             row_idx = table->add_empty_row();
-            table->set_string(m_schema.idx_identity, row_idx, identity);
-            table->set_bool(m_schema.idx_user_is_admin, row_idx, false);
-            m_realm->commit_transaction();
+            std::string uuid = util::uuid_string();
+            table->set_string(schema.idx_identity, row_idx, identity);
+            table->set_string(schema.idx_auth_server_url, row_idx, url);
+            table->set_string(schema.idx_local_uuid, row_idx, uuid);
+            table->set_bool(schema.idx_user_is_admin, row_idx, false);
+            table->set_bool(schema.idx_marked_for_removal, row_idx, false);
+            realm->commit_transaction();
         } else {
             // Someone beat us to adding this user.
-            m_realm->cancel_transaction();
+            row_idx = maybe_row_expr->get_index();
```

Comment:
Why get a row index from the `RowExpr` only to then retrieve a `Row` for it, rather than going directly to a `Row`?